python/osrd_schemas/scripts/truncate_infra.py

```diff
@@ -0,0 +1,378 @@
+    new_buffer_stops = []
+    former_switch_id_to_buffer_stop_id = {}
+    for switch in switches:
+        n_ports_included = sum(
+            port.track in track_section_ids for port in switch.ports.values()
+        )
+        if n_ports_included == len(switch.ports):
+            # All ports are included, no need to change anything
+            filtered_switches.append(switch)
+        elif n_ports_included > 1:
+            print(
+                f"Switch {switch.id} should be included in the bbox, but isn't handled yet"
+            )
+        elif n_ports_included == 1:
+            # Only one port is included, we need to replace the switch by a buffer stop
+            the_port = next(
+                port
+                for port in switch.ports.values()
+                if port.track in track_section_ids
+            )
+            new_buffer_stops.append(
+                BufferStop(
+                    id=f"buffer_stop_{switch.id}",
+                    track=the_port.track,
+                    position=0
+                    if the_port.endpoint == Endpoint.BEGIN
+                    else track_length_by_track[the_port.track],
+                )
+            )
+            former_switch_id_to_buffer_stop_id[switch.id] = (
+                new_buffer_stops[-1].id,
+                Direction.START_TO_STOP
+                if the_port.endpoint == Endpoint.BEGIN
+                else Direction.STOP_TO_START,
+            )
+
+    return filtered_switches, new_buffer_stops, former_switch_id_to_buffer_stop_id
+
+
+def filter_routes(
+    new_infra: RailJsonInfra,
+    old_infra: RailJsonInfra,
+    former_switch_id_to_buffer_stop_id: dict[str, tuple[str, Direction]],
+) -> list[Route]:
+    """Keep routes fully included in the bbox and adapt the ones that are partially included."""
+    waypoints_ids_kept = {b.id for b in new_infra.buffer_stops} | {
+        d.id for d in new_infra.detectors
+    }
+    switch_ids_kept = {s.id for s in new_infra.switches}
+
+    detectors_by_id = {d.id: d for d in old_infra.detectors}
+    buffer_stops_by_id = {b.id: b for b in old_infra.buffer_stops}
+    switches_by_id = {s.id: s for s in old_infra.switches}
+
+    res = []
+    for route in old_infra.routes:
+        entry_included = route.entry_point.id in waypoints_ids_kept
+        exit_included = route.exit_point.id in waypoints_ids_kept
+
+        n_switches_included = sum(
+            switch_id in switch_ids_kept
+            for switch_id in route.switches_directions.keys()
+        )
+
+        if entry_included and exit_included:
+            if n_switches_included == len(route.switches_directions):
+                # The route is fully included, no need to change anything
+                res.append(route)
+            else:
+                # This is an edge case not yet handled
+                print(f"Warning: dropping {route.id} that is partially included")
+
+        elif entry_included or exit_included:
```

Comment:
nitpick: it’s technically a XOR (the OR does work because you already treated the AND above, but yet, if code moves, making this one a XOR right now might be better).

```suggestion
        elif entry_included ^ exit_included:
```

Reply:
Done in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/f1fed4994d927d88a530c3b059ee8e59a8528e63.